components/authz-service/engine/opa/opa_internal_test.go

```diff
@@ -549,3 +429,864 @@ func randomPolicies(i int) []map[string]interface{} {
+	}
+
+	policies, roles := v2BaselinePoliciesAndRoles()
+
+	s.v2p1Store = inmem.NewFromObject(map[string]interface{}{
+		"policies": policies,
+		"roles":    roles,
+	})
+
+	teamCount := []int{0, 1, 10, 30, 50, 100, 150, 300, 500, 1000, 10000}
+	for _, count := range teamCount {
+		b.Run(fmt.Sprintf("V2FilterAuthorizedPairs with real life input and %d teams", count), func(b *testing.B) {
+			var resp []engine.Pair
+			var err error
+			for n := 0; n < b.N; n++ {
+				resp, err = s.V2FilterAuthorizedPairs(ctx, append([]string{"user:local:test@example.com"}, randomTeams(count)...), pairs)
+				if err != nil {
+					b.Error(err)
+				}
+			}
+			filteredPairsResp = resp
+		})
+	}
+}
+
+// Q: Which type of input is computed faster, generic Go interface or specific OPA Term?
+func BenchmarkV2GenericInput(b *testing.B) {
+	var r ast.Value
+	var err error
+	for n := 0; n < b.N; n++ {
+		// always record the result to prevent the compiler eliminating the function
+		// call.
+		r, err = genericV2Input(
+			[]string{"user:local:alice@example.com", "team:local:admins"},
+			"some:resource",
+			"some:resource:action",
+			[]string{"project-1", "project-2"})
+		if err != nil {
+			b.Fatal(err)
+		}
+	}
+	// always store the result to a package level variable
+	// so the compiler cannot eliminate the Benchmark itself.
+	result = r
+}
+
+func BenchmarkV2SpecificInput(b *testing.B) {
+	var r ast.Value
+	subjects, projects := []string{"user:local:alice@example.com", "team:local:admins"}, []string{"project-1", "project-2"}
+	resource, action := "some:resource", "some:resource:action"
+	for n := 0; n < b.N; n++ {
+		subs := make(ast.Array, len(subjects))
+		for i, sub := range subjects {
+			subs[i] = ast.NewTerm(ast.String(sub))
+		}
+		projs := make(ast.Array, len(projects))
+		for j, proj := range projects {
+			projs[j] = ast.NewTerm(ast.String(proj))
+		}
+		r = ast.NewObject(
+			[2]*ast.Term{ast.NewTerm(ast.String("subjects")), ast.NewTerm(subs)},
+			[2]*ast.Term{ast.NewTerm(ast.String("resource")), ast.NewTerm(ast.String(resource))},
+			[2]*ast.Term{ast.NewTerm(ast.String("action")), ast.NewTerm(ast.String(action))},
+			[2]*ast.Term{ast.NewTerm(ast.String("projects")), ast.NewTerm(projs)},
+		)
+	}
+	result = r
+}
+
+// A: Specific input is faster!
+// BenchmarkV2GenericInput-8   	    328531	      4034 ns/op	     1664 B/op	    50 allocs/op
+// BenchmarkV2SpecificInput-8   	  572641	      2073 ns/op	     976 B/op	      30 allocs/op
+
+func BenchmarkV2AuthorizedProjectPreparedQueryWithIncreasingPolicies(b *testing.B) {
+	var r error
+	ctx := context.Background()
+
+	l, err := logger.NewLogger("text", "debug")
+	require.NoError(b, err, "init logger")
+	s, err := New(ctx, l)
+	require.NoError(b, err, "init state")
+
+	policyCount := []int{0, 5, 10, 20, 50, 100, 200, 1000}
+	roleCount := 10 // keep this constant while increasing policyCount
+
+	chefPolicies, _ := v2BaselinePoliciesAndRoles()
+
+	for _, count := range policyCount {
```

Comment:
Wondering if renaming as shown might make this a tad more clear...?
```suggestion
	for _, policyCount := range policyDataPoints {
```

Reply:
I'd rather keep `count` in the name for consistency, but I can make it more specific with `policy`